### PR TITLE
Add option to write file from file-like object and use in download_file

### LINKED ISSUE #3455: downloading of large files fails with urllib.request with recent Python 3.x
In `download_file` we use `urllib.request`, which seems to throw an error in `url_fd.read()` when trying to read the whole file at once.

It can be reproduced with this script:
```python
import urllib.request

x = urllib.request.urlopen('https://developer.download.nvidia.com/compute/cuda/11.0.2/local_installers/cuda_11.0.2_450.51.05_linux.run')
x.read()
```

```
Traceback (most recent call last):
  File "./foo.py", line 5, in <module>
    x.read()
  File "/usr/lib/python3.8/http/client.py", line 467, in read
    s = self._safe_read(self.length)
  File "/usr/lib/python3.8/http/client.py", line 608, in _safe_read
    data = self.fp.read(amt)
  File "/usr/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
  File "/usr/lib/python3.8/ssl.py", line 1241, in recv_into
    return self.read(nbytes, buffer)
  File "/usr/lib/python3.8/ssl.py", line 1099, in read
    return self._sslobj.read(len, buffer)
OverflowError: signed integer is greater than maximum
```

While `urllib` is bugged like this, we need to either read it in chunks ourselves or skip the naive `read` and combine it with the subsequent write to file via `shutil.copyfileobj`:
```python
with open('/dev/shm/out.raw', 'wb') as fh:
        shutil.copyfileobj(x, fh)
```

As a bonus, we won't be reading the whole thing into memory, possibly exhausting it.

### PR DESCRIPTION
**Pull request by Flamefire**
Fixes downloads of large files where the size exceeds the maximum integer size
Fixes #3455

### CODE REVIEW

**Review by boegel — CHANGES_REQUESTED** [review]
see https://github.com/Flamefire/easybuild-framework/pull/6

File: easybuild/tools/filetools.py [hunk 1]
```diff
@@ -246,15 +246,20 @@ def write_file(path, data, append=False, forced=False, backup=False, always_over
...
+    if sys.version_info[0] >= 3 and (isinstance(data, bytes) or is_file_like):
         mode += 'b'
 
     # note: we can't use try-except-finally, because Python 2.4 doesn't support it as a single block
     try:
         mkdir(os.path.dirname(path), parents=True)
         with open_file(path, mode) as fh:
-            fh.write(data)
+            if is_file_like:
+                shutil.copyfileobj(data, fh)
```
[the remark this answers is not in the document]
Comment (Flamefire) [reply easybuild/tools/filetools.py:260]: IMO that is the task of the caller. Maybe not copying all is wanted. Moreover not all file-likes support seek, e.g. IIRC the response from the requests lib is a stream, so what was read is gone

File: easybuild/tools/filetools.py [hunk 2]
```diff
@@ -246,15 +246,20 @@ def write_file(path, data, append=False, forced=False, backup=False, always_over
     # cfr. https://docs.python.org/3/library/functions.html#open
     mode = 'a' if append else 'w'
 
+    is_file_like = hasattr(data, 'read')
```
[the remark this answers is not in the document]
Comment (Flamefire) [reply easybuild/tools/filetools.py:249]: > it's not unlikely something gets passed that has a read method

Do you have an example in mind?

I wanted to limit the check to what is required, and all that is required for the `copyfileobj` is a read method, nothing more.

I think the only other 2 possible arguments are string and bytes, everything else wouldn't be accepted by the `write` method anyway.
Speaking of that: `IOBase` doesn't provide `write`: https://docs.python.org/3/library/io.html#class-hierarchy

[the remark this answers is not in the document]
Comment (boegel) [reply easybuild/tools/filetools.py:249]: OK, makes sense.

File: easybuild/tools/filetools.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (boegel) [reply easybuild/tools/filetools.py:260]: Good point.

**Review by boegel — APPROVED** [review]

**Comment by boegel** [comment]
Thanks a lot for looking into this fix @Flamefire!

**Pull request merged**